hail/src/main/scala/is/hail/expr/ir/Binds.scala

```diff
@@ -15,10 +15,16 @@ object Bindings {
   // A call to Bindings(x, i) may only query the types of children with
   // index < i
   def apply(x: BaseIR, i: Int): Iterable[(String, Type)] = x match {
-    case Let(bindings, _) =>
-      val result = Array.ofDim[(String, Type)](i)
-      for (k <- 0 until i) result(k) = bindings(k)._1 -> bindings(k)._2.typ
-      result
+    case Block(bindings, _) =>
+      val result = Array.newBuilder[(String, Type)]
+      result.sizeHint(i)
+      val scope = if (i < bindings.length) bindings(i).scope else Scope.EVAL
```

Comment:
That is what I first wrote, which is wrong. The reason why is hard to see, because it's scattered across this file. I'm working on a separate change to consolidate the logic in this file, so all the details about a given node are in one place, which I think I'll try to rebase this on top of to clarify these changes.

To see why this is wrong, we have to look at these pieces:
```scala
object UsesAggEnv {
  def apply(ir0: BaseIR, i: Int): Boolean = ir0 match {
    case AggLet(_, _, _, false) => i == 0
}

object AggBindings {
  def apply(x: BaseIR, i: Int, parent: BindingEnv[_]): Option[Iterable[(String, Type)]] = x match {
      case AggLet(name, value, _, false) =>
        if (i == 1) wrapped(FastSeq(name -> value.typ)) else None
  }
}

object ChildEnvWithoutBindings {
  def apply[T](ir: BaseIR, i: Int, env: BindingEnv[T]): BindingEnv[T] = ir match {
      case _: IR =>
        if (UsesAggEnv(ir, i)) env.promoteAgg else if (UsesScanEnv(ir, i)) env.promoteScan else env
  }
}

object NewBindings {
  def apply(x: BaseIR, i: Int, parent: BindingEnv[_]): BindingEnv[Type] =
    BindingEnv(
      Env.fromSeq(Bindings(x, i)),
      agg = AggBindings(x, i, parent).map(b => Env.fromSeq(b)),
      scan = ScanBindings(x, i, parent).map(b => Env.fromSeq(b)),
      relational = Env.fromSeq(RelationalBindings(x, i)),
    )
}

object ChildBindings {
  def apply(ir: BaseIR, i: Int, baseEnv: BindingEnv[Type]): BindingEnv[Type] = {
    val env = ChildEnvWithoutBindings(ir, i, baseEnv)
    val newBindings = NewBindings(ir, i, env)
    env.merge(newBindings)
}
```

So in the old `AggLet`, the value uses the agg environment, meaning the agg environment is "promoted", replacing the eval environment, in `ChildEnvWithoutBindings`. That means in a chain of `AggLet`s
```
AggLet("a", a, AggLet("b", b, AggLet("c", c, ...)))
```
all previous bindings are in the eval environment in the value of each later binding, because the previous were added to the agg environment, which was then promoted to eval in the value.


In the new `Bind` node, this chain gets collapsed. Conceptually, in the value of an agg binding, previous agg bindings add to the agg environment, which is the promoted to the eval environment. But the combined effect is that previous agg bindings are added to the eval environment in an agg binding.

Reply:
This should now be much clearer after rebasing on #14404